AK/String.cpp

```diff
@@ -0,0 +1,338 @@
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Checked.h>
+#include <AK/Format.h>
+#include <AK/Memory.h>
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <AK/Utf8View.h>
+#include <stdlib.h>
+
+namespace AK {
+
+class StringData final : public RefCounted<StringData> {
+public:
+    static ErrorOr<NonnullRefPtr<StringData>> create_uninitialized(size_t, u8*& buffer);
+    static ErrorOr<NonnullRefPtr<StringData>> create_substring(StringData const& superstring, size_t start, size_t byte_count);
+    static ErrorOr<NonnullRefPtr<StringData>> from_utf8(char const* utf8_bytes, size_t);
+
+    struct SubstringData {
+        StringData const* superstring { nullptr };
+        u32 start_offset { 0 };
+    };
+
+    void operator delete(void* ptr);
+
+    ~StringData();
+
+    SubstringData const& substring_data() const
+    {
+        return *reinterpret_cast<SubstringData const*>(m_bytes_or_substring_data);
+    }
+
+    // NOTE: There is no guarantee about null-termination.
+    ReadonlyBytes bytes() const
+    {
+        if (m_substring) {
+            auto const& data = substring_data();
+            return data.superstring->bytes().slice(data.start_offset, m_byte_count);
+        }
+        return { &m_bytes_or_substring_data[0], m_byte_count };
+    }
+
+    StringView bytes_as_string_view() const { return { bytes() }; }
+
+    bool operator==(StringData const& other) const
+    {
+        return bytes_as_string_view() == other.bytes_as_string_view();
+    }
+
+    unsigned hash() const
+    {
+        if (!m_has_hash)
+            compute_hash();
+        return m_hash;
+    }
+
+private:
+    explicit StringData(size_t byte_count);
+    StringData(StringData const& superstring, size_t start, size_t byte_count);
+
+    void compute_hash() const;
+
+    u32 m_byte_count { 0 };
+    mutable unsigned m_hash { 0 };
+    mutable bool m_has_hash { false };
+    bool m_substring { false };
+
+    u8 m_bytes_or_substring_data[0];
+};
+
+String::String(NonnullRefPtr<StringData> data)
+    : m_data(&data.leak_ref())
+{
+}
+
+String::String(ShortString short_string)
+    : m_short_string(short_string)
+{
+}
+
+String::String(String const& other)
+    : m_data(other.m_data)
+{
+    if (!is_short_string())
+        m_data->ref();
+}
+
+String::String(String&& other)
+    : m_data(exchange(other.m_data, nullptr))
+{
+}
+
+String& String::operator=(String&& other)
+{
+    m_data = exchange(other.m_data, nullptr);
+    return *this;
+}
+
+String& String::operator=(String const& other)
+{
+    if (&other != this) {
+        m_data = other.m_data;
+        if (!is_short_string())
+            m_data->ref();
+    }
+    return *this;
+}
+
+String::~String()
+{
+    if (!is_short_string() && m_data)
+        m_data->unref();
+}
+
+String::String()
+{
+    m_short_string.byte_count_and_type = 0x01;
+}
+
+ErrorOr<String> String::from_utf8(StringView view)
+{
+    if (view.length() <= MAX_SHORT_STRING_LENGTH) {
```

Comment:
Let's not repeat the same mistakes we're tring to fix here :)

`MAX_SHORT_STRING_BYTE_COUNT`